File: python/context.py
#!/usr/bin/env python3
import ast
import argparse
import sys
import os
import glob
import re
import itertools


IGNORE_DIRECTORIES = ["__pycache__", ".git"]
SEARCH_DEFAULT, SEARCH_LINENO, SEARCH_REGEX, SEARCH_DEFINITIONS = range(4)


def echo(*args):
    """A poor man's print"""
    output = " ".join([str(i) for i in args]) + "\n"
    sys.stdout.write(output)


class SourceCode(object):
    def __init__(self, filename, offset=1):
        """
        offset is the difference between the number of the first line of the file and 0.  It's probably 1.
        """
        self.offset = offset
        self.filename = filename
        self.lines = []
        with open(filename, "r") as f:
            self.lines = f.readlines()

    def __repr__(self):
        return "SourceCode('{}', {})".format(self.filename, self.offset)

    def line(self, lineno):
        return self.lines[lineno - self.offset]

    @property
    def numlines(self):
        return len(self.lines)

    def format_line(self, lineno):
        return "{}:  {}".format(str(lineno).rjust(len(str(self.numlines))), self.line(lineno))


def make_matcher(search_type, look_for):
    def _regex(node):
        try: lineno = node.lineno
        except: return None
        for attr in node._fields:
            try:
                if look_for.match(getattr(node, attr)):
                    return lineno
            except:
                pass

    def _default(node):
        try: lineno = node.lineno
        except: return None
        for attr in node._fields:
            if getattr(node, attr) == look_for:
                return lineno       

    def _lineno(node):
        try: lineno = node.lineno
        except: return None
        return lineno if lineno == look_for else None

    # Convert look_for to the type that the matcher functions need.
    # Doing it here rather than inside the matcher function because it only needs to happen once
    look_for = {SEARCH_DEFAULT: lambda x: x,
                SEARCH_LINENO: int,
                SEARCH_REGEX: re.compile}[search_type](look_for)
    return {SEARCH_DEFAULT: _default,
            SEARCH_LINENO: _lineno,
            SEARCH_REGEX: _regex}[search_type]


def walk(node, matcher, history=None):
    history = [] if history is None else history[:]
    matches = []

    def append_to_history(item):
        try: history.append(item.lineno)
        except AttributeError: pass

    def pop_from_history():
        try: history.pop()
        except IndexError: pass

    children = list(ast.iter_child_nodes(node))
    match = matcher(node)

    if match:
        matches.append(match)
        matches = list(set(matches + history))

    if len(children) == 0:
        return set(matches)

    append_to_history(node)
    for i in children:
        append_to_history(i)
        matches.extend(walk(i, matcher, history))
        pop_from_history()
    pop_from_history()

    return set(matches)


def find_top_level(tree, depth=1):
    matches = [i.lineno for i in ast.iter_child_nodes(tree) if isinstance(i, (ast.FunctionDef, ast.ClassDef))]
    return matches


def find_context(tree, look_for, matcher):
    """
    Return a list of lines indicating branches that lead to the object we're looking for
    """
    # TODO: Add something to parse look_for, for example FooClass.bar should only match bar
    #       function calls that are attributes of a FooClass instance
    matches = walk(tree, matcher)
    return sorted(list(matches))


def parse_source(filename):
    with open(filename, "r") as f:
        tree = ast.parse(f.read(), filename=filename)
    return tree


def main(look_for, files, search_type=SEARCH_DEFAULT, recursive=False, ignore=IGNORE_DIRECTORIES, verbose=False):
    """
    look_for is a string, files is a list of paths
    """
    if recursive:
        files = itertools.chain(*[os.walk(f) for f in files])
    else:
        # format returned by os.walk:
        files = [(os.curdir, [], [f]) for f in files]

    all_contexts = {}
    skipped_files = {}

    for (directory, _, filenames) in files:
        abspath = os.path.abspath(directory)

        current_ignore = []
        for i in ignore:
            current_ignore.extend(glob.glob(os.path.abspath(i)))

        if abspath in current_ignore:
            continue
        for fn in filenames:
            source_file = os.path.join(directory, fn)
            if os.path.abspath(source_file) in current_ignore or source_file in current_ignore:
                continue
            try:
                source = SourceCode(source_file)
                tree = parse_source(source_file)
                if search_type == SEARCH_DEFINITIONS:
                    context = find_top_level(tree)
                else:
                    context = find_context(tree, look_for, make_matcher(search_type, look_for))
            except KeyboardInterrupt as e:
                sys.exit(1)
            except Exception as e:
                skipped_files[source_file] = "{}: {}".format(e.__class__.__name__, e)
                continue
            if len(context) == 0:
                continue

            if recursive or len(files) > 1:
                echo("\n{}".format(source_file))

            all_contexts[source_file] = context
            echo("\n" + "".join([source.format_line(i) for i in context]))

    if verbose and skipped_files:
        echo("Skipped these files due to errors:\n{}".format(
            "\n".join(["{}: {}".format(key, skipped_files[key]) for key in skipped_files])))

    # to make testing easier, don't have to capture and parse stdout:
    return all_contexts


if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        description="Find lines in a Python source file and the context they're in."
    )
    parser.add_argument("look_for", help="object to look for in the file")
    parser.add_argument("paths", help="files (or directories if -r option) to look in", nargs="+")
    parser.add_argument("-r", "--recursive", action="store_true", help="recursively search directory")
    parser.add_argument("-n", "--search-line", dest="search_type", action="store_const", const=SEARCH_LINENO,
                        help="search by line number")
    parser.add_argument("-e", "--search-regex", dest="search_type", action="store_const", const=SEARCH_REGEX,
                        help="search by regexp")
    parser.add_argument("-d", "--search-defs", dest="search_type", action="store_const", const=SEARCH_DEFINITIONS,
                        help=("just look for class and function definitions.  The look_for argument "
                              "should be an integer indicating the maxmimum depth of search"))
    parser.add_argument("-v", "--verbose", action="store_true",
                        help="display information about errors and skipped files")
    parser.add_argument("-i", "--ignore",
                        help=("comma-separated list of files and directories "
                              "to ignore, default is {}".format(IGNORE_DIRECTORIES)))
    args = parser.parse_args()

    if args.ignore:
        ignore = args.ignore.split(",") + IGNORE_DIRECTORIES
    else:
        ignore = IGNORE_DIRECTORIES

    main(args.look_for,
         args.paths,
         search_type=args.search_type or SEARCH_DEFAULT,
         recursive=args.recursive,
         ignore=ignore,
         verbose=args.verbose)

File: ruby/context.rb
#!/usr/bin/env ruby
require 'ripper'
require 'stringio'
require 'pp'
require 'optparse'
require 'find'

oldstderr, $stderr = $stderr, StringIO.new  # suppress warnings
require 'parser/current'
$stderr = oldstderr


SEARCH_DEFAULT, SEARCH_LINENO, SEARCH_REGEX, SEARCH_DEFINITIONS = (0..3).to_a
IGNORE_DIRECTORIES = [".git"]
BLUE = "\e[34m"
LIGHT_GREEN =  "\e[92m"
END_COLOR = "\e[0m"


class SourceCode
  def initialize(filename, look_for: nil, num_color: nil, line_color: nil, offset: 1)
    @filename = filename
    @offset = offset
    @lines = File.readlines(filename)
    @look_for = look_for
    @num_color = num_color || ""
    @line_color = line_color || ""
    @end_color = (num_color || line_color) ? END_COLOR : ""
  end
  
  def add_line_color(line)
    #formatted_lookfor = make_formatter(format).call(@look_for)
    #string.gsub(Regexp.new(@look_for), formatted_lookfor)
    colored_look_for = @line_color + @look_for + @end_color
    line.gsub(Regexp.new(@look_for), colored_look_for)
  end

  def add_num_color(lineno)
    @num_color + lineno + @end_color
  end
  
  def line(lineno)
    @lines[lineno - @offset]
  end

  def numlines
    @lines.length
  end

  def format_line(lineno)
    colored_lineno = add_num_color( lineno.to_s.rjust(numlines.to_s.length) )
    colored_line = add_line_color( line(lineno) )
    "#{colored_lineno}:  #{colored_line}"
  end
end


def make_matcher(search_type, look_for)
  if search_type == SEARCH_REGEX
    look_for = Regexp.new(look_for)
  elsif search_type == SEARCH_LINENO
    look_for = look_for.to_i
  end

  _default = lambda do |node, lineno|
    if node.respond_to?(:children) && node.children.map(&:to_s).include?(look_for)
      return lineno
    end
  end

  _regex = lambda do |node, lineno|
    node._fields.each do |attr|
      begin
        return lineno if look_for.match(getattr(node, attr))
      rescue
      end
    end
  end

  _lineno = lambda do |node, lineno|
    return lineno == look_for ? lineno : nil
  end

  {SEARCH_DEFAULT => _default,
   SEARCH_LINENO => _lineno,
   SEARCH_REGEX => _regex}[search_type]
end


def walk(node, matcher, history=nil)
  history = history.nil? ? [] : history.dup.uniq
  matches = []

  lineno = node.location.line rescue nil
  match = matcher.call(node, lineno)
  if match
    matches.push(match)
    matches = (matches + history).uniq
  end
 
  if !node.is_a?(AST::Node) || node.children.nil? || node.children.length == 0
    return matches.uniq
  end

  history.push(lineno)
  node.children.each do |child|
    history.push(child.location.line) rescue nil
    matches += walk(child, matcher, history)
    history.pop
  end
  history.pop

  return matches.compact.sort.uniq
end

  
def find_top_level(tree, depth=1)
end


def main(look_for, files, search_type: SEARCH_DEFAULT, recursive: false,
         ignore: IGNORE_DIRECTORIES, verbose: false, color: false)
  num_color = color ? BLUE : nil
  line_color = color ? LIGHT_GREEN : nil

  # check for recursive, walk files/directories
  all_contexts = {}
  skipped_files = {}

  if recursive
    files = files.map{ |dir|
      Find.find(dir).map { |f|
        File.expand_path(f, Dir.getwd)
      }
    }.flatten
  end

  files.each do |f|
    source_file = f  # grab absolute path here
    context = []
    begin
      source = SourceCode.new(source_file, look_for: look_for, num_color: num_color, line_color: line_color)
      tree = Parser::CurrentRuby.parse_file(source_file)

      if search_type == SEARCH_DEFINITIONS
        context = find_top_level(tree)
      else
        context = walk(tree, make_matcher(search_type, look_for))
      end
    rescue SystemExit, Interrupt
      raise
    rescue => e
      skipped_files[f] = e
    end
    next if context.length == 0

    puts source_file if recursive || files.length > 1

    context.each { |lineno| puts source.format_line(lineno) }
    puts ""
  end

  if verbose && skipped_files.length > 0
    puts "Skipped these files due to errors:"
    skipped_files.each { |k, v| puts "#{k}:  #{v}" }
  end
  return all_contexts
end


if __FILE__ == $0
  options = {}
  OptionParser.new do |opts|
    opts.set_banner("Find lines in a Ruby source file and the context they're in\n\n" +
                    "Usage: #{$0} [options] look_for paths [paths...]\n\n" +
                    "Options:\n")
    opts.on("-v", "--verbose", "display information about errors and skipped files") { |v| options[:verbose] = true }
    opts.on("-r", "--recursive", "recursively search directory") { |v| options[:recursive] = true }
    opts.on("-n", "--search-line", "search by line number") { |v| options[:search_type] = SEARCH_LINENO }
    opts.on("-e", "--search-regex", "search by regular expression") { |v| options[:search_type] = SEARCH_REGEX }
    opts.on("-c", "--color", "highlight line numbers and matches") { |v| options[:color] = true }
    opts.on("-d", "--search-defs", "look for class, module, and function definitions") { |v|
      options[:search_type] = SEARCH_DEFINITIONS
    }
    opts.on("-h", "--help", "print this help") { puts opts; exit }
    opts.on("-i IGNORE", "--ignore IGNORE", "comma-separated list of files and directories to ignore") { |v|
      options[:ignore] = IGNORE_DIRECTORIES + v.split(",")
    }
    if ARGV[0].nil? || ARGV[1].nil?
      puts opts; exit
    end
  end.parse!

  look_for = ARGV[0]
  paths = ARGV.slice(1, ARGV.length)

  main(look_for, paths,
       search_type: options[:search_type] || SEARCH_DEFAULT,
       recursive: options[:recursive],
       ignore: options[:ignore],
       verbose: options[:verbose],
       color: options[:color])
end
